style-guide: clean up "must"/"should"/"may" (#113380)
rust-lang/rust: src/doc/style-guide/src/expressions.md

### Chains of fields and method calls

A chain is a sequence of field accesses, method calls, and/or uses of the try
operator `?`. E.g., `a.b.c().d` or `foo?.bar().baz?`.

> Prefer formatting on one line if possible, and the chain is *small*. If

Reviewer: Should we drop "prefer" here too in the spirit of the language clean up?

Perhaps something like
```suggestion
Format the chain one line if it is *small* and otherwise possible to do so. If
```

Author: I kept some of the "prefer" where they seemed to represent "here's how to choose among multiple possible sources of line breaks", but this one I agree is unconditional.

}
```

> If the control line needs to be broken, prefer to break before the `=` in `*

Reviewer: Similar thought as earlier wrt to "prefer", but don't feel strongly

```suggestion
If the control line needs to be broken, break before the `=` in `*
```

Author: Here, I *think* that this "prefer" represents "here's how to choose among multiple possible sources of line breaks". For instance, if you *have* to break it in another place because this break isn't sufficient, and with it broken in that other place you don't need to break before the `=`, then you shouldn't actually break it before the `=`.

If we want to say "always break before the `=` if you have to break at all, even if when done you don't need that break to stay under the line width", we could make this change, but I don't think that's accurate here?

### Match

> Prefer not to line-break inside the discriminant expression. Always break after

Reviewer: Recognize this is another instance of existing "prefer" language. I don't think it's worth trying to change it now, but I'm curious how operative this really is. Whether the expression needs to be broken should be dependent on the respective rules for that expression, unless there's some match specific context carveout i'm nt familiar with

Author: @calebcartwright I *think* this is a semantically meaningful "prefer": if you need to break the line, and you have multiple places you could do so, try others first, and only break within the discriminant if nothing else suffices to get it below the line length.

)];
```

> Apply this behavior recursively.

Reviewer: Recognize this is part of removing a "may" and references to tools, though I'm slightly paranoid that there may be cases where rustfmt used the existing language to break the recursion which may no longer be correct with the proposed changes.

I'm not quite sure how to handle this one. Thoughts?

For a function with multiple arguments, if the last argument is a multi-line
closure with an explicit block, there are no other closure arguments, and all
the arguments and the first line of the closure fit on the first line, use the

> same combining behavior:

Reviewer: This is _much_ clearer now :+1:

Do not put spaces in ranges, e.g., `0..10`, `x..=y`, `..x.len()`, `foo..`.

When writing a range with both upper and lower bounds, if the line must be
broken within the range, break before the range operator and block indent the
second line:

```rust
a_long_expression
    ..another_long_expression
```

For the sake of indicating precedence, if either bound is a compound
expression, use parentheses around it, e.g., `..(x + 1)`, `(x.f)..(x.f.len())`,
or `0..(x - 10)`.

### Hexadecimal literals

Hexadecimal literals may use upper- or lower-case letters, but they must not be

> mixed within the same literal. Projects should use the same case for all

Reviewer: I assume this is an intentional case of leaving "should"?

Author: Yeah, unfortunately. I thought about just removing this language entirely, but that'd be a semantic change, so I figured if I did that at all it could be in a separate PR with an FCP.
